extensions/aws/processors/PutKinesisStream.cpp

```diff
@@ -0,0 +1,164 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "PutKinesisStream.h"
+
+#include <memory>
+#include <random>
+#include <unordered_map>
+
+#include "aws/kinesis/KinesisClient.h"
+#include "aws/kinesis/model/PutRecordsRequest.h"
+#include "core/ProcessContext.h"
+#include "core/ProcessSession.h"
+#include "core/Resource.h"
+#include "utils/ProcessorConfigUtils.h"
+
+namespace org::apache::nifi::minifi::aws::processors {
+
+void PutKinesisStream::initialize() {
+  setSupportedProperties(Properties);
+  setSupportedRelationships(Relationships);
+}
+
+void PutKinesisStream::onSchedule(core::ProcessContext& context, core::ProcessSessionFactory& session_factory) {
+  AwsProcessor::onSchedule(context, session_factory);
+
+  batch_size_ = parseU64Property(context, MessageBatchSize);
+  if (batch_size_ == 0 || batch_size_ > 500) {
+    logger_->log_warn("PutKinesisStream::MessageBatchSize is invalid. Setting it to the maximum 500 value.");
+    batch_size_ = 500;
+  }
+  batch_data_size_soft_cap_ = parseDataSizeProperty(context, MaxBatchDataSize);
+  if (batch_data_size_soft_cap_ > 4_MB) {
+    logger_->log_warn("PutKinesisStream::MaxMessageBufferSize is invalid. Setting it to the maximum 4 MB value.");
+    batch_data_size_soft_cap_ = 4_MB;
+  }
+
+  endpoint_override_url_ = context.getProperty(EndpointOverrideURL.name) | minifi::utils::toOptional();
+}
+
+struct StreamBatch {
+  uint64_t batch_size = 0;
+  std::vector<std::shared_ptr<core::FlowFile>> flow_files;
+};
+
+void PutKinesisStream::onTrigger(core::ProcessContext& context, core::ProcessSession& session) {
+  logger_->log_trace("PutKinesisStream onTrigger");
+
+  constexpr uint64_t SINGLE_RECORD_MAX_SIZE = 1_MB;
+  std::unordered_map<std::string, StreamBatch> stream_batches;
+  auto credentials = getAWSCredentials(context, nullptr);
+
+  if (!credentials) {
+    logger_->log_error("Failed to get credentials for PutKinesisStream");
+    context.yield();
+    return;
+  }
+
+  for (uint64_t i = 0; i < batch_size_; i++) {
+    std::shared_ptr<core::FlowFile> flow_file = session.get();
+    if (!flow_file) { break; }
+    const auto flow_file_size = flow_file->getSize();
+    if (flow_file_size > SINGLE_RECORD_MAX_SIZE) {
+      flow_file->setAttribute(AwsKinesisErrorMessage.name, fmt::format("record too big {}, max allowed {}", flow_file_size, SINGLE_RECORD_MAX_SIZE));
+      session.transfer(flow_file, Failure);
+      logger_->log_error("Failed to publish to kinesis record {} because the size was greater than {} bytes", flow_file->getUUID().to_string(), SINGLE_RECORD_MAX_SIZE);
+      continue;
+    }
+
+    auto stream_name = context.getProperty(AmazonKinesisStreamName.name, flow_file.get());
+    if (!stream_name) {
+      logger_->log_error("Stream name is invalid due to {}", stream_name.error().message());
+      session.transfer(flow_file, Failure);
+      continue;
+    }
```

Comment:
we could set the `aws.kinesis.error.message` attribute here, too

Reply:
good idea, https://github.com/apache/nifi-minifi-cpp/pull/1927/commits/3bf6cb8601b1584cdb9ee1c9a252ee97577699da#diff-7f440937a3123a187fcf52fd1ab61e0d4617345e57a49cbff318988e18cf24d2R90